### ISSUE TITLE
Add support for JIT compilation

### ISSUE THREAD

**Issue by blythed** [lf1-io/padl#222]
Work has begun in this branch https://github.com/lf1-io/padl/tree/issue_222

**Comment by wuhu** [comment]
How about using https://pytorch.org/docs/1.10.0/generated/torch.jit.ignore.html?highlight=jit%20ignore#torch.jit.ignore?
We could, in the init of TorchModuleTransform apply it to all methods coming from Transform.

I had a brief look at the branch, what is the purpose of `pd_to_torchscript`? I'm not sure we need it.
Also we should think about how to save/load torchscript modules, perhaps we need to implement a new set of `pre_save`, `post_load`.

**Comment by sjrl** [comment]
I was thinking `pd_to_torchscript` could be a useful function to have to `jit` `script` all `TorchModuleTransforms` contained within a `padl.Pipeline`. Otherwise the user would need to do it for each `TorchModuleTransform` individually. What do you think? This might be useful if you train the `padl.Pipeline` using regular `torch.nn.Module` and only after training would like to make a `jit` `script` version. This is pretty common due to the lack of feature support for `jit` `script` compared to `torch.nn.Module`'s such as checkpointing.

**Comment by sjrl** [comment]
> How about using https://pytorch.org/docs/1.10.0/generated/torch.jit.ignore.html?highlight=jit%20ignore#torch.jit.ignore? We could, in the init of TorchModuleTransform apply it to all methods coming from Transform.

That looks really promising. I think that would be worth doing instead of my hacky suggestion. 

> Also we should think about how to save/load torchscript modules, perhaps we need to implement a new set of `pre_save`, `post_load`.

Yes, especially given the above it looks like we won't be able to just call the `jit_model.save(file_name)` due to the `@torch.jit.ignore` decorator. In this case we will want to save the  `jit` `script` in addition to the model weights, right?

**Comment by sjrl** [comment]
> How about using https://pytorch.org/docs/1.10.0/generated/torch.jit.ignore.html?highlight=jit%20ignore#torch.jit.ignore? We could, in the init of TorchModuleTransform apply it to all methods coming from Transform.

I started playing with this but unfortunately it looks like `torch.jit.ignore` is incompatible with `@property` shown in this issue https://github.com/pytorch/pytorch/issues/54688. Alternatively `torch.jit.unused` works with `@property`, but that property would no longer be accessible (e.g. `torch.jit.script(MyModuleTransform).pd_layers` would raise an error.

**Comment by wuhu** [comment]
Ok, that's annoying. Only way then would be to get rid of all `property`s? 
Or we try what you suggested. I'm not 100% sure of the implications though, possibly this could break something. I don't know enough about torchscript to be sure if this would work.

About `pd_to_torchscript` - agree that sounds useful. Still not sure if I would put it into core padl as it seems to be more of an utility. How about moving it to another issue and concentrating on just enabling jit.script support here? That way we determine what to do with it later.

**Comment by sjrl** [comment]
> Ok, that's annoying. Only way then would be to get rid of all `property`s? Or we try what you suggested. I'm not 100% sure of the implications though, possibly this could break something. I don't know enough about torchscript to be sure if this would work.

I agree that my idea doesn't sound bullet proof. I'll try removing all `property`'s first. This shouldn't affect performance with `lru_cache` since that works just fine with functions and I think `torch.jit.ignore` will work it as well. 

> About `pd_to_torchscript` - agree that sounds useful. Still not sure if I would put it into core padl as it seems to be more of an utility. How about moving it to another issue and concentrating on just enabling jit.script support here? That way we determine what to do with it later.

Okay sounds good I can move it to a separate issue. I don't think the code is too difficult, but I agree we need to answer where it should go.

**Comment by sjrl** [comment]
@wuhu I'm finding that this is becoming more complicated and I think I need some help. We ultimately have two scenarios that we are interested in supporting 
```python
class NN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(16, 8)
        
    def forward(self, x):
        return self.linear(x)
```
## 1. Transform wrapped around `torch.jit.script`
```python
jit_nn = transform(torch.jit.script(NN()))
```
This actually seems to already work including saving (you can see tests for this in the branch https://github.com/lf1-io/padl/tree/issue_222)

## 2. `torch.jit.script` wrapped around `transform`
This scenario is important because users may want to convert their `torch.nn.Module` code to `jit.script` after training. 
```python
jit_nn = torch.jit.script(transform(NN()))
```
* This requires us to wrap all `padl` related functions by `torch.jit.ignore`, which has been done in the above issue_222 branch
* Additionally, we cannot use the `@property` decorator ([pytorch issue](https://github.com/pytorch/pytorch/issues/54688)), which has also been done in the above issue_222 branch
**Remaining Issues:** 
* At this point `jit_nn` contains most of the padl functions and Transform attributes. However, the type of `jit_nn` is no longer a `Transform`, which means that `jit_nn` is missing the class attributes `pd_mode` and `_pd_external_full_dump_modules`. 
* Additionally, all of the other attributes have become jit scripted as well meaning they are statically typed. For example, `jit_nn._pd_device = None` throws the error `RuntimeError: Could not cast attribute '_pd_device' to type str: Unable to cast Python instance to C++ type (compile in debug mode for details)`. This may not cause problems, but it is something to be aware of. 

I'm not currently sure how to overcome these issues.

**Comment by sjrl** [comment]
Perhaps the best way to solve situation 2 above is to provide a convenience function like `pd_to_torch_script` which under the hood runs something like situation 1, using the saved source code for the original `torch.nn.Module`. What do you think? @wuhu
